dl/translation: fixes for OOM handling

bharathv opened the pull request:
Main changes in the PR

* Adds exceptional handling to all abortable paths in the file writer and propagate them all the way into translation context via appropriate error codes.
* In abortable paths (eg: scheduler time limit exceeded/ OOM), the output is gracefully handled and flushed instead of discarding to make forward progress.
* Handles OOM handling for pathological cases by adding reservation per writer. 
* Added tests that exercised these exceptional paths and found a few relevant bugs.



<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
dl/dt: Pathological partitioning test

Basically maps each record to its own partition. Nodes OOM in the multiplexor

bharathv committed:
dl/translation: introduce oom_error ec

Used to represent a translation task failure due to memory exhaustion.

bharathv committed:
dl/translation/writer: propagate an abort source to create writer

Creation of a writer will involve a reservation and hence needs to be
abortable.

bharathv committed:
dl/translation: refactor translation mem tracking

Two main changes

- Centralizes mem tracking in translation_ctx. This memory tracker is
  shared by all parquet writers from the same translator to reserve
  memory. Having access to the memory tracker within translation_ctx
  makes the call to buffered_bytes() inexpensive (doesn't need to loop
  through every writer to compute buffered bytes as they are rolled up
  into the mem tracker)

- Changes API to do reserve(bytes, as) and free(bytes, as). The existing
  API turned out to be buggy in the presence of multiple writers as we
  overwrite current usage from the latest writer. This change is also a
  preparatory work for the next commit that wires up a fixed cost per
  writer.

bharathv committed:
dl/translation/writer: factor in a fixed size overhead per writer

bharathv committed:
dl/translation: propagate abort errors all the way

Disambiguates various translation codepaths with their specific
exception types and corresponding writer_error codes. This makes most
data writer path noexcept and appropriate error codes are populated.

This behavior is used in the next commits to help make partition
translator forward progress if translator runs into these preemption
code paths.

bharathv committed:
dl/translator: graceful handling of OOM/scheduling timeouts

This commit classifies the writer errors into two classes recoverable
and non recoverable.

Recoverable errors are the ones caused because the translation was
purposefully preempted by the scheduler due to time/memory violation.
Every other class of errors and considered unrecoverable.

Recoverable errors does not leave the writers in a bad shape, the
translator can just flush them and make progress if it wishes to, which
happens in the case of memory limit being exceeded.

bharathv commented:
/dt

bharathv committed:
dt: add test for exceptional code paths

bharathv commented:
/dt

bharathv committed:
dl/translation: add ability to translate from the middle of a batch

Translation can be aborted in the middle of a batch which is then
flushed to the coordinator and the subsequent iteration should start
from the later offset. Since log reader only operates on batch
boundaries we need to layer this additional check in the multiplexer to
skip already translated offsets (by passing an additional param).

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/63324-->
<!--0c30bb66-d724-44d7-95ab-1e9f97a3deae-->
<!--317c0d29-c1ea-4203-96ac-804d60c7303d-->
<!--title-->
## Retry command for [Build#63324](https://buildkite.com/redpanda/redpanda/builds/63324)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/translation_interruption_test.py::DatalakeTranslationInterruptionsTest.test_scheduler_time_slice_interruptions@{"catalog_type":"rest_hadoop","cloud_storage_type":1,"query_engine":"spark"}
tests/rptest/tests/datalake/custom_partitioning_test.py::DatalakeCustomPartitioningTest.test_many_partitions@{"catalog_type":"nessie","cloud_storage_type":1}
```
<!--/body-->

bharathv committed:
dl/multiplexer: populate result offsets after add_data succeeds

add_data may fail (for example writer exceeded memory budget) in which
case we should _not_ update result offsets. It should only be done
_after_ a write succeeds which guarantees its persistence.

bharathv commented:
Could use a little bit more polish (hence the draft), but overall ready for review.

bharathv commented:
/dt

bharathv commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -137,32 +137,6 @@ partition_translator::checkpoint_translation_result(
       });
 }
 
-ss::future<>
-partition_translator::translate_when_notified(kafka::offset begin_offset) {
-    co_await _ready_to_translate.wait(
-      [this] { return _inflight_translation_state.has_value(); });
-
-    auto& as = _inflight_translation_state->as;
-    auto reader = co_await _data_source->make_log_reader(
-      begin_offset, datalake_priority(), as);
-    if (!reader) {
-        co_return;
-    }
-    vlog(_logger.trace, "starting translation from offset: {}", begin_offset);
-    ss::timer<scheduling::clock> cancellation_timer;
-    cancellation_timer.set_callback([&as] { as.request_abort(); });
-
-    auto translation_f
-      = _translation_ctx
-          ->translate_now(
-            std::move(reader.value()), _inflight_translation_state->as)
-          .finally([this] { return _translation_ctx->flush(); });
-    cancellation_timer.arm(_inflight_translation_state->translate_for);
-    co_await std::move(translation_f).finally([&cancellation_timer] {
-        cancellation_timer.cancel();
-    });
-}
-
```
dead code, forgot to remove from a previous refactor, nothing to review here.

bharathv commented on src/v/datalake/record_multiplexer.cc:
```diff
@@ -94,17 +94,19 @@ record_multiplexer::record_multiplexer(
 
 ss::future<> record_multiplexer::multiplex(
   model::record_batch_reader reader,
+  kafka::offset start_offset,
```
I vaguely remember seeing a similar behavior in some existing tests in some rare cases, this bug was caught by the interruption tests.

andrwng replied:
Nice catch

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--df2131f6-df1b-4858-8a33-f4a74bf08074-->
<!--TEST RESULTS-->
<!--33c60d24-5676-4a19-9c33-19d2733d00b3-->
<!--TEST RESULTS-->
<!--9fd5eb7d-88f8-472d-a12b-157d16c47a7f-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63330">build#63330</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_record_multiplexer_rpfixture.gtest_record_multiplexer_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/63330#0195acb5-bb37-4604-84e4-fcefce82d24b | FAIL | 0/2 |
| gtest_record_multiplexer_rpfixture.gtest_record_multiplexer_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/63330#0195acb5-bb38-449b-a0d8-1daafa3c0c94 | FAIL | 0/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63330#0195acf6-755f-4c6d-a8ad-d520ee4dde2c | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63330#0195acfa-b9fb-4e17-a08d-e57b46eb31a5 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63330#0195acf6-7560-4316-a71e-8b469867b68d | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/63330#0195acf6-755e-48f3-bed1-4c9240d1955a | FLAKY | 1/2 |
| rptest.tests.datalake.rest_catalog_connection_test.RestCatalogConnectionTest.test_redpanda_connection_to_rest_catalog.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/63330#0195acf6-755f-4c6d-a8ad-d520ee4dde2c | FLAKY | 1/2 |
| rptest.tests.tiered_storage_pause_test.TestTieredStoragePause.test_safe_pause_resume.allow_gaps_topic_level=True.allow_gaps_cluster_level=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/63330#0195acfa-b9fc-484c-a3c4-e3c497916a81 | FLAKY | 1/2 |
| rptest.tests.timequery_test.TimeQueryKafkaTest.test_timequery_below_start_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/63330#0195acfa-b9fc-484c-a3c4-e3c497916a81 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63435">build#63435</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/63435#0195b3f3-644a-433b-9eed-c720b0d36204 | FLAKY | 1/2 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape | https://buildkite.com/redpanda/redpanda/builds/63435#0195b3f3-644a-433b-9eed-c720b0d36204 | FLAKY | 1/3 |
| rptest.tests.datalake.datalake_upgrade_test.DatalakeUpgradeTest.test_upload_through_upgrade.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK | ducktape | https://buildkite.com/redpanda/redpanda/builds/63435#0195b40e-3695-4cae-90c9-8244c38fc796 | FLAKY | 1/2 |
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recommissioning_node_finishes | ducktape | https://buildkite.com/redpanda/redpanda/builds/63435#0195b3f3-6449-4b65-bca5-d66fd91bad80 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/63471">build#63471</a></summary>


| test_id                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture                                                                                                                                       | unit        | https://buildkite.com/redpanda/redpanda/builds/63471#0195b626-0558-4940-90b2-1f3e36e00760 | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63471#0195b667-e54b-46f7-940f-e285681922ca | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=False.params=.cancellation.dir.in.stage.preparing.use_alias.True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63471#0195b667-e549-48ee-a74a-d1912646e7a5 | FLAKY         | 1/2      |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_migrated_topic_data_integrity.transfer_leadership=True.params=.cancellation.dir.in.stage.preparing.use_alias.False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63471#0195b667-e54b-46f7-940f-e285681922ca | FLAKY         | 1/2      |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63471#0195b667-e54d-4e77-ac9a-0c93729cecfe | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/63471#0195b66b-5510-44b4-bb49-9e425844f84e | FLAKY         | 1/4      |
</details>
<!--/body-->

bharathv commented:
/dt

andrwng commented on src/v/datalake/serde_parquet_writer.cc:
```diff
@@ -20,9 +20,16 @@ ss::future<writer_error> serde_parquet_writer::add_data_struct(
+        } else if (new_buffered_bytes < _buffered_bytes) {
+            co_await _mem_tracker.free_bytes(
+              _buffered_bytes - new_buffered_bytes, as);
+        }
```
nit: chatted on slack about when this could happen -- would be nice to comment that we may get here because the underlying writers will transparently apply compression at some byte interval. It can otherwise be surprising for writing a row to result in less buffered memory

bharathv replied:
added a comment.

andrwng commented on src/v/datalake/serde_parquet_writer.h:
```diff
@@ -25,8 +25,8 @@ class serde_parquet_writer : public parquet_ostream {
     writer_mem_tracker& _mem_tracker;
-    size_t _buffered_bytes{0};
-    size_t _flushed_bytes{0};
+    int64_t _buffered_bytes{0};
+    int64_t _flushed_bytes{0};
```
Can these ever be negative? Or would that be an accounting bug?

Also, was this change necessary? It doesn't look like any other types changed and it looks like they're only set as stats.buffered_size and stats.flushed_size

bharathv replied:
clang complained otherwise

```
Comparison of integers of different signs: 'int64_t' (aka 'long') and 'size_t' (aka 'unsigned long')clang(-Wsign-compare)

```

for

```
new_buffered_bytes > _buffered_bytes
```

writer tracks them as int64_t, I think because arrow library does so (or java long is signed), practically they are never negative.

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -491,13 +497,15 @@ class partition_translation_context : public translation_context {
 
     ss::future<checked<coordinator::translated_offset_range, translation_errc>>
     finish(retry_chain_node& rcn, ss::abort_source& as) final {
+        _mem_tracker.release();
```
Would it make sense to only release the memory when exiting this function?

andrwng replied:
Bump?

bharathv replied:
done. ideally its not needed because flush() is always called after every iteration, added a comment and moved it to end of this block.

andrwng commented on src/v/datalake/local_parquet_file_writer.cc:
```diff
@@ -175,7 +176,17 @@ local_parquet_file_writer_factory::local_parquet_file_writer_factory(
+    // Additionally one other contributor per writer is the buffer used
+    // in the output stream which defaults to 8_KiB, which is only released
+    // on output stream close().
+    static constexpr size_t WRITER_RESERVATION_OVERHEAD = 10_KiB;
```
Curious if this is something we can measure in a test, or is it something we have to empirically figure out (and keep track of in case the underlying data structures change?)

bharathv replied:
ya this seemed like a conservative upper bound to begin with, definitely needs empirical evaluation and adjustment. Also depends on the shape of partition key schema and such I think (since they are encoded into the writer metadata).


added a TODO.

andrwng commented on src/v/datalake/translation/deps.cc:
```diff
@@ -485,6 +485,16 @@ class partition_translation_context : public translation_context {
 
     ss::future<> flush() final {
         if (_in_progress_translation) {
+            // Note: The flush here *does not* fully release memory associated
```
nit: feels like this should be a TODO rather than just a note

bharathv replied:
done

andrwng commented on src/v/datalake/data_writer_interface.cc:
```diff
@@ -26,10 +26,29 @@ std::ostream& operator<<(std::ostream& os, const writer_error& ev) {
+    case time_quota_exceeded:
+        return writer_error::time_limit_exceeded;
+    case unknown:
+        return writer_error::shutting_down;
```
nit: is this actually the closest thing it will map to? If not, I worry a bit about callers seeing shutting_down and then assuming the entire system is shutting down and e.g. stop all further translations or something, vs some generic flush_error or somesuch

bharathv replied:
thats a fair point, remapped.

andrwng replied:
Still seeing this as shutting_down?

bharathv replied:
eeks done! brain fade, changed in all other places except this mapping.

andrwng commented on src/v/datalake/translation/partition_translator.h:
```diff
@@ -167,7 +167,7 @@ class partition_translator : public scheduling::translator {
     fetch_translation_offsets(retry_chain_node&);
 
-    ss::future<>
+    ss::future<bool>
```
nit: add a comment about what it returns? Or bool_class?

bharathv replied:
done.

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -284,13 +259,22 @@ ss::future<> partition_translator::run_one_translation_iteration(
-          "Translation attempt failed: {}, discarding state to reset "
-          "translation",
-          ex);
+          "Translation attempt ran into OOM, result will be flushed "
```
nit: just for peace of mind i wonder if it's worth calling this something other than OOM, like it hit the reservation limit or something. OOM sounds a little scary? (not sure 🤔)

bharathv replied:
sure, reworded.

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -284,13 +259,22 @@ ss::future<> partition_translator::run_one_translation_iteration(
           _logger.warn,
-          "Translation attempt failed: {}, discarding state to reset "
-          "translation",
-          ex);
+          "Translation attempt ran into OOM, result will be flushed "
+          "immediately");
+        force_flush_after_iteration = true;
+    } catch (const translator_time_quota_exceeded_error&) {
+        vlog(
+          _logger.debug,
+          "Translation attempt exceeded scheduler time limit quota");
```
nit: maybe add an explicitly comment that we're not keeping the exceptions because the underlying state is still safe to use. Could be mistaken for an oversight

bharathv replied:
done.

andrwng reviewed (commented):
Nice, high level this makes sense to me. I'm a little hesitant about this new recoverable category of errors because it means we need to think about getting there at any point in translation, but conceptually it feels like the right thing to do

bharathv reviewed (commented):
thanks for the review

bharathv commented:
Had to rebase to fix conflicts, removed some uneeded custom_partitioning test commits

bharathv commented:
/dt

bharathv commented:
/dt

andrwng commented on src/v/datalake/local_parquet_file_writer.cc:
```diff
@@ -195,7 +195,11 @@ local_parquet_file_writer_factory::create_writer(
     // sizing.
     static constexpr size_t WRITER_RESERVATION_OVERHEAD = 10_KiB;
-    co_await _mem_tracker.reserve_bytes(WRITER_RESERVATION_OVERHEAD, as);
+    auto reservation_err = co_await _mem_tracker.reserve_bytes(
```
nit: if this is no longer a draft it'd be nice if each non-trivial commit had a meaningful commit message

bharathv replied:
done, this hunk was misplaced, moved to previous commit.

andrwng commented on src/v/datalake/tests/record_multiplexer_test.cc:
```diff
@@ -158,7 +158,13 @@ class RecordMultiplexerTestBase
             auto reader = model::make_memory_record_batch_reader(
               std::move(subset));
-            mux.multiplex(std::move(reader), model::no_timeout, as).get();
+            mux
```
Worth adding a simple unit test? I guess the interruptions test caught this, but it seems like a surprising way to catch this bug

andrwng commented on src/v/datalake/record_multiplexer.cc:
```diff
@@ -287,31 +287,30 @@ ss::future<ss::stop_iteration> record_multiplexer::do_multiplex(
-
         auto& writer = writer_iter->second;
-        auto write_result = co_await writer->add_data(
+        auto add_data_result = co_await writer->add_data(
```
nit: commit message?

piyushredpanda commented:
/backport v25.1

piyushredpanda commented:
/backport v25.1.x

vbotbuildovich commented:
Oops! Something went wrong.

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13984743451).

vbotbuildovich commented:
Failed to create a backport PR to v25.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25423-v25.1.x-627 remotes/upstream/v25.1.x
git cherry-pick -x ef1ba6e014 0bf9b1d2a9 bdd24b13f9 69106dc821 7d339a8896 d433adac1f 6dc1950bc7 5fcb56dd22 32d3548cae 92126bb711
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13984744745).
